File: leetcode/sum-root-to-leaf-numbers.cpp
// https://leetcode.com/problems/sum-root-to-leaf-numbers/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int sumNumbers(TreeNode* root) {
        return work(root, 0);
    }
    
    int work(TreeNode *node, int val) {
        if (!node)
            return 0;
        int next = 10 * val + node->val;
        if (!node->left && !node->right)
            return next;
        return work(node->left, next)
             + work(node->right, next);
    }
};

File: leetcode/sum-root-to-leaf-numbers.go
// https://leetcode.com/problems/sum-root-to-leaf-numbers/

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func sumNumbers(root *TreeNode) int {
    return work(root, 0)
}

func work(node *TreeNode, val int) int {
    if node == nil {
        return 0
    }
    next := 10 * val + node.Val
    if node.Left == nil && node.Right == nil {
        return next
    }
    return work(node.Left, next) + work(node.Right, next)
}
